### ISSUE TITLE
Nuclide run-through

### ISSUE THREAD

**Issue by vjeux** [jlongster/jscodefmt#2]
This looks awesome already. I ran the formatter through the nuclide codebase ( https://github.com/facebook/nuclide/tree/master/pkg ) and here are the things I'd love to get tweaked so that we can use it at Facebook :)

- [x] Make last block expand

![image](https://cloud.githubusercontent.com/assets/197597/21577201/0aea3098-cf53-11e6-97e3-2f72e56a3a72.png)

When the last argument of a function call is a function, object or array, and all the previous arguments fit in one line, want to inline it.

```js
call(a, b, () => {

});
```

- [x] Put one argument per line when it expands

![image](https://cloud.githubusercontent.com/assets/197597/21577210/5af22e56-cf53-11e6-9e5a-a3e0d4ab1161.png)

- [x] Do not put spaces inside of `{}` and `[]`

![image](https://cloud.githubusercontent.com/assets/197597/21577213/72d7616c-cf53-11e6-9ab8-0c819356db1f.png)

- [x] Use `,` for type object separator instead of `;`

![image](https://cloud.githubusercontent.com/assets/197597/21577217/87382a56-cf53-11e6-89e7-846b0e880018.png)

- [x] Prioritize breaking on `?` and `:`

![image](https://cloud.githubusercontent.com/assets/197597/21577221/aaaf38bc-cf53-11e6-8593-bb2f8d381d39.png)

- [x] Use single quote

![image](https://cloud.githubusercontent.com/assets/197597/21577223/bf127ed6-cf53-11e6-8102-442a26684009.png)

- [x] Add trailing commas

![image](https://cloud.githubusercontent.com/assets/197597/21577226/c9090c02-cf53-11e6-94be-cd1ef3d8e650.png)

- [x] Preserve original \n

![image](https://cloud.githubusercontent.com/assets/197597/21577229/e786770a-cf53-11e6-980b-e6fef69c820e.png)

I don't think it's possible to automatically add empty lines in the right place most of the time. I think this should be under the developer control to decide where to put them.

- [x] Do not put an extra indent for else if

![image](https://cloud.githubusercontent.com/assets/197597/21577239/4e707290-cf54-11e6-8ee9-682800131b2e.png)

- [ ] Invalid code generated

![image](https://cloud.githubusercontent.com/assets/197597/21577248/87fc8468-cf54-11e6-9034-e7a5cb8c4f00.png)

- [x] Remove trailing whitespace

![image](https://cloud.githubusercontent.com/assets/197597/21577252/9d4c5762-cf54-11e6-87a8-61c7c7196874.png)

- [x] Implement chaining for `&` and `|`

![image](https://cloud.githubusercontent.com/assets/197597/21577258/e575703c-cf54-11e6-8b8d-48d712435897.png)

- [x] Implement call chains

![image](https://cloud.githubusercontent.com/assets/197597/21577266/24714676-cf55-11e6-880b-bddf8d123ccf.png)

```
Something
  .call(
    ...
  )
  .something.call( ... )
  .call( ... );
```

The way I want it to look is that if it doesn't fit into a single line, every .call() is on its own line and indented two spaces. The way to do it is to look for CallExpression and traverse all the CallExpressions and MemberExpressions and rewrite them in this way.

**Comment by jlongster** [comment]
All of these look really good! None of them changes the few things I'm opinionated about, so I'm really happy to work with you on this. Most of these are pretty easy. Also, I think we should have CLI options to configure a few of them (singe/double-quotes and spaces inside objects/arrays at least). That will help adoption.

> Prioritize breaking on ? and :

It will be interesting to figure this out because the way it formats it now is because of how the AST is structured, which is based on operator precedence. If we force it to be different, we'll probably run into other weird formatting when the ternary operator is involved in a longer expression. We should create a test file with a bunch of different types of expressions involving a ternary.

I have no problems changing this though, hopefully it's not too hard.

> I don't think it's possible to automatically add empty lines in the right place most of the time. I think this should be under the developer control to decide where to put them.

Yeah, I think you're right. FWIW refmt does not do this, and I was trying to be as opinionated as it. However, I think this would be another thing that would help adoption.

I think we could at least collapse multiple empty lines into a single empty line. Otherwise, keep single and double newlines in the output, which allows the programmer to group code.

> Invalid code generated

That's not invalid! :) I'm kidding, I know what you mean. It moved the comments up. Comments are really hard, and that behavior is from recast, because it doesn't know if the comment is attached to the 


> Remove trailing whitespace



> Implement call chains

**Comment by vjeux** [comment]
Cool!

For collapsing more than a single empty line, I'm onboard :)

**Comment by jlongster** [comment]
Should have more time tomorrow to fix most of these; already finished two. Does the function args change look good?

https://github.com/jlongster/jscodefmt/commit/fa16988401fb62e27418d197ef32fabd9fc60406

**Comment by jlongster** [comment]
Specializing last argument expansion is finished in https://github.com/jlongster/jscodefmt/commit/588c8ce7cda33e4fe58a004cf524bebf7934014d (and a follow-up commit, there are a few errors in those snapshot changes).

**Comment by vjeux** [comment]
<3

**Comment by jlongster** [comment]
No more spaces inside objects/arrays in https://github.com/jlongster/jscodefmt/commit/8326963f2d7a8ef5014957ca861a1d319bea1a0b. You can configure this by passing the `bracketSpacing` option (`jscodefmt --bracket-spacing` will turn it on).

**Comment by jlongster** [comment]
Uses `,` for separating flow type fields in https://github.com/jlongster/jscodefmt/commit/b550f1d088ccab19f9b1db08014e4a9a898f1903

**Comment by jlongster** [comment]
Single quotes can be enabled in https://github.com/jlongster/jscodefmt/commit/599b4311bb6be9204689a7725d37d6fdfca770aa with the `--single-quotes` option or the `quote: "single"` API option (default to double for now).

**Comment by jlongster** [comment]
@vjeux A few questions about trailing commas:

1. Should they only be printed if the expression is broken across newlines? So it would print the following:

```js
// No comma
{ one: 1, two: 2 }

// Has comma
{
  one: 1,
  two: 2,
}
```

2. Which things should print trailing commas out of the following?
  * ObjectExpressions
  * ArrayExpressions
  * Function args (trailing comma is illegal, right?)
  * Flow tuple/object types

**Comment by jlongster** [comment]
In https://github.com/jlongster/jscodefmt/commit/fc186c8df8d2117db406a316ecc476ab1925a705 you can add trailing commas to objects split across lines with the `--trailing-comma` option. If there are other expressions that should have a trailing comma, let me know.

**Comment by vjeux** [comment]
@jlongster for commas:

That's correct, only print trailing comma when they are split in groups.

For function args, we have a tc39 proposal at stage 4 ("Finished: Indicate that the addition is ready for inclusion in the formal ECMAScript standard") https://github.com/tc39/proposal-trailing-function-commas and a babel transform that we've been using for years https://babeljs.io/docs/plugins/syntax-trailing-function-commas/

We want everything that is separated by commas to have trailing commas. All the ones you mentioned are √. @pieterv probably has an exhaustive list.

**Comment by jlongster** [comment]
> We want everything that is separated by commas to have trailing commas. All the ones you mentioned are √. @pieterv probably has an exhaustive list.

Hm... I'm not sure I want to print trailing commas in a place that require a babel transform. I suppose I could do that, since by default it doesn't print trailing commas, and if you opt in you get it everywhere. If users complain maybe we'll have to provide a slightly more fine-grain option.

**Comment by jlongster** [comment]
else if's should be combined in https://github.com/jlongster/jscodefmt/commit/fcd5436b383266f38e49f75d1f67dbd666f13a82

**Comment by jlongster** [comment]
@vjeux Alright, in https://github.com/jlongster/jscodefmt/commit/0d4d21cc68ab63b3696e07f19d311fa7652a4c74 it prints trailing commas everywhere. If I missed any place, let me know.

**Comment by jlongster** [comment]
Ternary operators will break across lines first in https://github.com/jlongster/jscodefmt/commit/7b0ec6d93546c2c1882df21a9edbf9f6d7c2a6da. This actually was easy; I just hadn't implemented any breaking behavior for ternary ops.

**Comment by jlongster** [comment]
Better rules for breaking in union/intersection types landed in https://github.com/jlongster/jscodefmt/commit/5c39dc02e62a8b202a32b6af15eed3ce21969bff

**Comment by vjeux** [comment]
Omg, you are on a streak <3

**Comment by jlongster** [comment]
Added call chains in https://github.com/jlongster/jscodefmt/pull/4

**Comment by jlongster** [comment]
Original blank lines are preserverd in https://github.com/jlongster/jscodefmt/commit/bcd44b4368ce90b013409efd11509c0df1dd7711. @vjeux I'd be interested in how the nuclide codebase looks now after that; does it work properly?

**Comment by jlongster** [comment]
I improved printing of comments in https://github.com/jlongster/jscodefmt/commit/994432c193c5318bfe38d3917b7754c632e4f20e. There is still probably more work to do there.

Unfortunately I could not solve this case:

```js
this._configCache = LRU({
  max: 2000, // Want this to exceed the maximum...
  maxAge: 1000 * 30, // 30 seconds
});
```

See the [issue on recast](https://github.com/benjamn/recast/issues/191#issuecomment-118104105) that I linked to before. It's very difficult because although it's attached as a trailing comment on the right object property AST node, that node is printed before the comma. Therefore a naive printing will do this:

```js
this._configCache = LRU({
  max: 2000 // Want this to exceed the maximum...,
  maxAge: 1000 * 30 // 30 seconds,
});
```

Note how the comma is printed after the comment. Fixing this would involve some really complicated logic for how to interlave comments.

But it's even harder for us than recast. Since we are a flexible printer, we will collapse nodes when they fix. That means that inline comments will be collapsed as well, and even if we fixed the comma, we would output this:

```js
this._configCache = LRU({ max: 2000, // Want this to ... maxAge: 1000 * 30 // 30 seconds })
```

Basically users cannot reliably have inline comments because even if we outputted them correctly, they might be collapsed. We could also try and force parents to break if they have inline comments, but all of this will be really hard.

For now, we do what recast does and simply convert those trailing comments to a leading comment which is much easier to print. We can look into improving this later, although once you apply our printer, it won't convert leading comments back to trailing ones...

**Comment by jlongster** [comment]
I think I have completed this issue as much as I can right now. Feel free to file more issues specific to comments so that we can track them. I'm sure there are other things we can do, but I'd like to go ahead and announce this to get more feedback.
